### ISSUE TITLE
Randomly failing SAMFire test

### ISSUE THREAD

**Issue by jan-car** [hyperspy/hyperspy#1522]
When trying to test HyperSpy with py.tests on my machine, I get error messages of the following type for all test cases:

```python
file hyperspy\conftest.py, line 13
  @pytest.fixture(autouse=True)
  def add_np(doctest_namespace):
        fixture 'doctest_namespace' not found
        available fixtures: recwarn, capfd, capsys, tmpdir_factory, cache, tmpdir, monkeypatch, add_np, record_xml_property, pytestconfig
        use 'py.test --fixtures [testpath]' for help on them.
```

I assume the solution is easy, but I'm not familliar with py.tests, at all. Do I have to use some specific keywords or options when executing the tests? I assume the `hyperspy\tests` folder is still the right target for the automatic test search, right?

What is the benefit of using py.tests instead of nosetest?

Thanks and sorry, if this is a stupid question :-).

(I'll tag @woozey, because I assume he is interested as well)

**Comment by to266** [comment]
`pytest` is maintained and simpler to write, among other things. You could argue it's "more modern", I suppose... It also supports parallel testing that we don't actually use because it clashes with our parallel/lazy tests, but it's there as well :)

I guess your pytest version is older? See doc [here](http://doc.pytest.org/en/latest/doctest.html#the-doctest-namespace-fixture)

I just run it as 
```bash
pytest hyperspy/tests
```
or just `pytest` in the hyperspy folder. It successfully discovers the tests and the required setup, but the lookup may be slightly slower

**Comment by jan-car** [comment]
Updating to 3.0 did indeed fix the issue!

I still have around 10 failed tests, but these are not due to pytests itself (e.g. I did not upgrade to matplotlib 2.0, yet, due to incompatibilities with PyQt4 which I still need).

Thanks a lot :-), I'll close the issue!

**Comment by ericpre** [comment]
With matplotlib < 2.0, you will have image comparison failure when you run `py.test --mpl`. Running `py.test` only will run the plotting code but not perform the image comparison, so in this case, it should not failed. What is the failure?

I am surprised about the incompatibility issue between matplotlib 2.0 and PyQt4. You should be able to use both together.

**Comment by jan-car** [comment]
Ah, I assumed matplotlib 2.0 was dependant on PyQt5, because conda always tried to update PyQt together with matplotlib. I managed to update matplotlib without changing PyQt now. Thanks @ericpre for pointing this out, I wanted to update for a bit now...

There are still 9 failed tests left:
- Two of them are (I assume) due to the `unbcf_fast` library not being present on my system. I tried to build it by hand, but apparently the c compiler can't be found. I should be able to sort this out, somehow...
- Then I have a TypeError in `test_blockfile.py` in line 217 in `test_load_readonly`. I read that `dask.array` does not support item assignements and I assume this function should test this, but it raises a `TypeError` instead of a `NotImplementedError` (which was expected by pytest, I assume?).
```python
with pytest.raises(NotImplementedError):
>           s.data[:] = 23
E           TypeError: 'Array' object does not support item assignment
```
- The next error occurs in line 105 of test_date_time_tools in test_update_date_time_in_metadata, which expects the `time_zone` metadata to be `'Coordinated Universal Time'`, but on my german system it is `'Koordinierte Weltzeit'`. Nothing drastic, but it will cause problems for other OS languages, I assume.
- The last five are all `KeyErrors` in lazy tests, e.g. in `test_integrate_in_range.py` in line 23:
```python
def test_integrate_in_range(self):
        integrated_signal = self.signal.integrate_in_range(signal_range=(None, None))
>       assert np.allclose(integrated_signal.data, 20,)
signal\test_integrate_in_range.py:23: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
C:\Users\Jan\Anaconda3\lib\site-packages\numpy\core\numeric.py:2478: in allclose
    res = all(isclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan))
C:\Users\Jan\Anaconda3\lib\site-packages\numpy\core\numeric.py:2548: in isclose
    x = array(a, copy=False, subok=True, ndmin=1)
C:\Users\Jan\Anaconda3\lib\site-packages\dask\array\core.py:898: in __array__
    x = self.compute()
C:\Users\Jan\Anaconda3\lib\site-packages\dask\base.py:86: in compute
    return compute(self, **kwargs)[0]
C:\Users\Jan\Anaconda3\lib\site-packages\dask\base.py:175: in compute
    for opt, val in groups.items()])
C:\Users\Jan\Anaconda3\lib\site-packages\dask\base.py:175: in <listcomp>
    for opt, val in groups.items()])
C:\Users\Jan\Anaconda3\lib\site-packages\dask\array\optimization.py:22: in optimize
    dsk2, dependencies = cull(dsk, keys)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

dsk = {('array-946e3825e554e085d547a7cd0b0d03ee', 0): (<function getarray at 0x00000000058D3C80>, 'array-original-946e3825e5...2.97343903e-06,   2.98834196e-06,   3.00331658e-06, ...,
         3.01836322e-06,   3.00331658e-06,   2.98834196e-06])}
keys = [('simps-8871a51f818b0c69babffd65b8879f58', 0)]

    def cull(dsk, keys):
        """ Return new dask with only the tasks required to calculate keys.
    
        In other words, remove unnecessary tasks from dask.
        ``keys`` may be a single key or list of keys.
    
        Examples
        --------
        >>> d = {'x': 1, 'y': (inc, 'x'), 'out': (add, 'x', 10)}
        >>> dsk, dependencies = cull(d, 'out')  # doctest: +SKIP
        >>> dsk  # doctest: +SKIP
        {'x': 1, 'out': (add, 'x', 10)}
        >>> dependencies  # doctest: +SKIP
        {'x': set(), 'out': set(['x'])}
    
        Returns
        -------
        dsk: culled dask graph
        dependencies: Dict mapping {key: [deps]}.  Useful side effect to accelerate
            other optimizations, notably fuse.
        """
        if not isinstance(keys, (list, set)):
            keys = [keys]
        out = dict()
        seen = set()
        dependencies = dict()
        stack = list(set(flatten(keys)))
        while stack:
            key = stack.pop()
>           out[key] = dsk[key]
E           KeyError: ('simps-8871a51f818b0c69babffd65b8879f58', 0)

C:\Users\Jan\Anaconda3\lib\site-packages\dask\optimize.py:43: KeyError
```

Sorry if this is just an info dump, but maybe it is of some significance, for someone else?

**Comment by jan-car** [comment]
I'm not sure if reopening is the correct procedure. Maybe I should open separate issues for the different error types, instead? Feel free to correct me here :-)

EDIT: Sorry for spam-opening and closing...

**Comment by magnunor** [comment]
Thanks for reporting this! Making the tests run nicely on every platform/install is important.

Slicing issues into smaller pieces usually make them more manageable, so your suggestion sounds good.

> The next error occurs in line 105 of test_date_time_tools in test_update_date_time_in_metadata, which expects the time_zone metadata to be 'Coordinated Universal Time', but on my german system it is 'Koordinierte Weltzeit'. Nothing drastic, but it will cause problems for other OS languages, I assume.

The tests shouldn't be specific to one locale, so this should be fixed.

With regards to the dask errors: I've had similar errors, so I'm guessing there needs to be some kind of version specifier for one or more of the dependencies.

**Comment by jan-car** [comment]
I separated the reports into three issues:
#1523 
#1524 
#1525 
I'll leave the issue with the `unbcf_fast` library here, as I assume this is a problem on my side. Maybe someone has a solution for this as well? I haven't "installed" HyperSpy properly as I simply cloned the GitHub repository to contribute to the development, but due to this, I did not install or even build `unbcf_fast`. Trying to build the extension via `setup.py`, I get:
```
setup.py:178: UserWarning: WARNING: C compiler can't be found.
Only slow pure python alternative functions will be available.
To use fast implementation of some functions writen in cython/c either:
a) check that you have compiler (EXACTLY SAME as your python
distribution was compiled with) installed,
b) use binary distribution of hyperspy (i.e. wheels, egg, (only osx and win)).
```
I assume installing via conda would work, but then I would have HyperSpy in two places (my development project folder and in the package library of python itself). Is there another way?

**Comment by jan-car** [comment]
Okay... I installed HyperSpy via conda and magically this solved most my problems, I had to manually copy the compiled `unbcf_fast` to my development version of HyperSpy, but everything went smooth.
In the course of the installation, dask was set to version `0.13.0`, which solved issues #1523 and #1525 and I'll close both.
The dateutil issue in #1524 remains.

I only thing that remains is an `AssertionError` in `test_samfire.py/test_multiprocessed`:

```python
number of pixels failed: 102
F
self = <hyperspy.tests.samfire.test_samfire.TestSamfireMain object at 0x0000000011AC9A58>

    @cleanup
    def test_multiprocessed(self):
        self.model.fit()
        samf = self.model.create_samfire(ipyparallel=False)
        samf.plot_every = np.nan
        samf.strategies[0].radii = 1.
        samf.strategies.remove(1)
        samf.optional_components = [self.model[2]]
        samf.start(fitter='mpfit', bounded=True)
        # let at most 3 pixels to fail randomly.
        fitmask = samf.metadata.marker == -np.ones(self.shape)
        print('number of pixels failed: {}'.format(
              np.prod(self.shape) - np.sum(fitmask)))
>       assert np.sum(fitmask) >= np.prod(self.shape) - 5
E       assert 3 >= (105 - 5)
E        +  where 3 = <function sum at 0x000000000390E510>(array([[ True,  True, False, False, False, False, False, False, False,\n        False, False, False, False, False, Fals...False, False, False, False, False, False, False, False,\n        False, False, False, False, False, False]], dtype=bool))
E        +    where <function sum at 0x000000000390E510> = np.sum
E        +  and   105 = <function prod at 0x000000000390EB70>((7, 15))
E        +    where <function prod at 0x000000000390EB70> = np.prod
E        +    and   (7, 15) = <hyperspy.tests.samfire.test_samfire.TestSamfireMain object at 0x0000000011AC9A58>.shape

samfire\test_samfire.py:330: AssertionError```

I assume this has nothing to do with the other errors, but I'm not sure...

**Comment by francisco-dlp** [comment]
That's an infamous randomly failing test.

**Comment by jan-car** [comment]
Ah, okay :-), should I close this issue then?

**Comment by francisco-dlp** [comment]
Not until the test is fixed ;)

**Comment by francisco-dlp** [comment]
I've updated the title to point at the remaining issue to fix in this thread.

**Comment by jan-car** [comment]
I can confirm the random nature of the error, in a run this evening, the test did not fail...

**Comment by francisco-dlp** [comment]
Fixed in #1569
